### PR TITLE
Enhanced pre/post test job hung/leftover process checking

### PR DESCRIPTION
**Pull request by andrew-m-leonard**
Fixes: https://github.com/adoptium/aqa-tests/issues/4184

Enhances the pre/post PROCESSCATCH logic checking for more left over processes, in particular:
- "make" : stops repetitive process killing at end of failed/aborted test jobs
- "jck" extended processes
- more specific "java" process selection

Enhanced to better filter Jenkins Agent process, and utilise Windows enhanced powershell queries.

### CODE REVIEW

File: buildenv/jenkins/JenkinsfileBase [hunk 1]
```diff
@@ -25,12 +25,12 @@ def makeTest(testParam) {
...
-	if (PLATFORM.contains("windows")) {
-		sh(script:'powershell -c "Get-Process java | select id,processname,starttime"', returnStatus:true)
-	} else {
-		sh(script:'ps -fu '+"${env.USER}"+' | grep java | egrep -v "remoting.jar|agent.jar|grep"', returnStatus: true)
+	// Terminate any previous test related processes that are running
+	echo "PROCESSCATCH: Terminating previous test related processes which are currently on the machine:"
+	def statusCode=sh(script:"aqa-tests/terminateTestProcesses.sh 2>&1", returnStatus:true)
+	echo "returnCode = ${statusCode}"
+	if (statusCode != 0) {
+		echo "Unable to terminate all processes."
```
Comment (llxia) [inline buildenv/jenkins/JenkinsfileBase:33]: Should we just call `terminateJavaProcesses()`?

Comment (andrew-m-leonard) [reply buildenv/jenkins/JenkinsfileBase:33]: yeah :-)

**Comment by andrew-m-leonard** [comment]
@llxia @sxa ready for review please

**Comment by llxia** [comment]
@AdamBrousseau could you also review this PR? Thanks

File: terminateTestProcesses.sh [hunk 2]
```diff
@@ -0,0 +1,84 @@
+#!/bin/sh
+
+# Arg S1 : Unix current env.User
+
+# If Windows use PowerShell
```
Comment (karianna) [inline terminateTestProcesses.sh:5]:
```suggestion
# If Windows uses PowerShell
```

File: buildenv/jenkins/JenkinsfileBase [hunk 3]
```diff
@@ -829,50 +824,14 @@ def testExecution() {
...
-	echo "PROCESSCATCH: Removed processes - here is what is remaining:"
-	if (PLATFORM.contains("windows")) {
-		sh(script:'powershell -c "Get-Process java | select id,processname,starttime"', returnStatus:true)
-	} else {
-		sh(script:'ps -fu '+"${env.USER}"+' | grep java | egrep -v "remoting.jar|agent.jar|grep"', returnStatus:true)
-	}
+        def statusCode=sh(script:"aqa-tests/terminateTestProcesses.sh ${env.USER} 2>&1", returnStatus:true)
+        echo "returnCode = ${statusCode}"
+        if (statusCode != 0) {
+                echo "Unable to terminate all processes."
```
Comment (karianna) [inline buildenv/jenkins/JenkinsfileBase:833]: Print out the error code here?

File: terminateTestProcesses.sh [hunk 4]
```diff
@@ -0,0 +1,84 @@
+#!/bin/sh
+
```
Comment (karianna) [inline terminateTestProcesses.sh:2]: set -ex ?

File: terminateTestProcesses.sh [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+# Arg S1 : Unix current env.User
+
+# If Windows use PowerShell
+if [ "$OSTYPE" = "cygwin" ]; then
+  echo "Windows machine, using powershell queries..."
+
+  # Match anything that most likely test job or process related
```
Comment (karianna) [inline terminateTestProcesses.sh:9]:
```suggestion
  # Match anything that is most likely a test job or process related
```

File: terminateTestProcesses.sh [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+             CommandLine like '%rmiregistry%' or \
+             CommandLine like '%tnameserv%' or \
+             CommandLine like '%make.exe%')"
+
+  # Ignore Jenkins agent and grep cmd
+  ignore_str="not CommandLine like '%remoting.jar%' and \
+              not CommandLine like '%agent.jar%' and \
+              not CommandLine like '%grep%'"
+
+  C=`powershell -c "(Get-WmiObject Win32_Process -Filter {${ignore_str} and ${match_str}} | measure).count" | tr -d "\\\\r"`
```
Comment (karianna) [inline terminateTestProcesses.sh:25]: What is a `C` :-)

File: terminateTestProcesses.sh [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  
+  if [ $C -gt 0 ]; then
+      echo Windows rogue processes detected, attempting to stop them..
+      powershell -c "Get-WmiObject Win32_Process -Filter {${ignore_str} and ${match_str}}"
+      powershell -c "(Get-WmiObject Win32_Process -Filter {${ignore_str} and ${match_str}}).Terminate()"
+      uname | grep CYGWIN 2>/dev/null || uptime
+      sleep 10
```
Comment (karianna) [inline terminateTestProcesses.sh:32]: Is that seconds?  Probably worth echoing that so users expect it

File: terminateTestProcesses.sh [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      uname | grep CYGWIN 2>/dev/null || uptime
+      C=`powershell -c "(Get-WmiObject Win32_Process -Filter {${ignore_str} and ${match_str}} | measure).count" | tr -d "\\\\r"`
+      if [ $C -gt 0 ]; then
+        echo "Cleanup failed, processes still remain..."
+        exit 127
+      fi 
+      echo "Processes stopped successfully"
+  else
+      echo Woohoo - no rogue processes detected
```
Comment (karianna) [inline terminateTestProcesses.sh:41]:
```suggestion
      echo Woohoo - no rogue processes detected!
```

File: terminateTestProcesses.sh [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  # Fix For Issue https://github.com/adoptium/infrastructure/issues/2442 - Solaris PS Command Truncation
+  if [ `uname` = "SunOS" ]; then
+      PSCOMMAND="/usr/ucb/ps -uxww"
+  else
+      PSCOMMAND="ps -fu $1"
+  fi
+
+  if $PSCOMMAND | egrep "${match_str}" | egrep -v "${ignore_str}"; then
+      echo Boooo - there are rogue processes kicking about
+      echo Issuing a kill to all processes shown above - though this will not terminate vfb processes on AIX as we are trying to debug source of leftovers there
```
Comment (karianna) [inline terminateTestProcesses.sh:63]: Seems very AIX specific, perhaps wrap in a detection block?

File: terminateTestProcesses.sh [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      $PSCOMMAND | egrep "${match_str}" | egrep -v "${ignore_str}" | awk '{print $2}' | xargs -n1 kill
+      uname | grep CYGWIN 2>/dev/null || uptime
+      sleep 10
```
Comment (karianna) [inline terminateTestProcesses.sh:66]: Same as above, consider echoing.

**Review by karianna — CHANGES_REQUESTED** [review]

File: terminateTestProcesses.sh (hunk 4 above)
Comment (andrew-m-leonard) [reply terminateTestProcesses.sh:2]: not here, don't want to exit at first failed shell command

File: terminateTestProcesses.sh (hunk 6 above)
Comment (andrew-m-leonard) [reply terminateTestProcesses.sh:25]: a count variable, i'll rename it!

File: terminateTestProcesses.sh (hunk 9 above)
Comment (andrew-m-leonard) [reply terminateTestProcesses.sh:63]: That's a copy&paste from somewhere else, i'll remove it

File: terminateTestProcesses.sh [hunk 11]
```diff
@@ -0,0 +1,83 @@
+#!/bin/sh
+
+# Arg S1 : Unix current env.User
+
+# If Windows uses PowerShell
+if [ "$OSTYPE" = "cygwin" ]; then
```
Comment (sxa) [inline terminateTestProcesses.sh:6]: It looks like this was `PLATFORM.contains("windows")` instead of an explicitly cygwin check previously. What is the reason for this change as there are alternative environments that can be used to build on Windows other than cygwin that would presumably not be covered after this.

File: terminateTestProcesses.sh [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  echo "Windows machine, using powershell queries..."
+
+  # Match anything that is most likely a test job or process related
+  match_str="(CommandLine like '%java%jck%' or \
+             CommandLine like '%openjdkbinary%java%' or \
+             CommandLine like '%java%javatest%' or \
+             CommandLine like '%java%-Xfuture%' or \
+             CommandLine like '%X%vfb%' or \
```
Comment (sxa) [inline terminateTestProcesses.sh:14]: `X%vfb` is a UNIX process, so should not be relevant on Windows.

File: terminateTestProcesses.sh [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+else
+  echo "Unix type machine.."
+
+  # Match anything that is most likely a jck test job or process related
+  match_str="java.*jck|openjdkbinary.*java|java.*javatest|java.*-Xfuture|X.*vfb|rmid|rmiregistry|tnameserv|make"
+
+  # Ignore Jenkins agent and grep cmd
+  ignore_str="remoting.jar|agent.jar|grep"
+
+  if [ `uname` = "SunOS" ]; then
```
Comment (sxa) [inline terminateTestProcesses.sh:53]: Note: This seems to have changed from the previous version. The new one will likely not work on Solaris 11.
(This code looks familiar ;-) I'm going to have to look separately on adjusting this at the source to be ok on Solaris 11!)

File: terminateTestProcesses.sh [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+             CommandLine like '%rmiregistry%' or \
+             CommandLine like '%tnameserv%' or \
+             CommandLine like '%make.exe%')"
+
+  # Ignore Jenkins agent and grep cmd
+  ignore_str="not CommandLine like '%remoting.jar%' and \
+              not CommandLine like '%agent.jar%' and \
+              not CommandLine like '%grep%'"
+
+  count=`powershell -c "(Get-WmiObject Win32_Process -Filter {${ignore_str} and ${match_str}} | measure).count" | tr -d "\\\\r"`
```
Comment (sxa) [inline terminateTestProcesses.sh:25]: Not sure why this has changed from the `GetProcess` used in the previous version of this function but this seems to be filtering on `CommandLine` which isn't the object returned by `Get-WmiObject Win32_Process`. Am I missing something?
image

**Review by sxa — CHANGES_REQUESTED** [review]
The Windows process check in particular is going to need checking, the others are suggestions and superfluous clauses.

File: terminateTestProcesses.sh (hunk 13 above)
Comment (andrew-m-leonard) [reply terminateTestProcesses.sh:52]: ok thanks?

Comment (andrew-m-leonard) [reply terminateTestProcesses.sh:52]: @sxa maybe other places that use uname SunOS check need changing? https://github.com/adoptium/aqa-tests/blob/d089897c5c9a7d87ba6def8efd53a4cb786c2e99/buildenv/jenkins/JenkinsfileBase#L425

File: terminateTestProcesses.sh (hunk 11 above)
Comment (sxa) [reply terminateTestProcesses.sh:6]: I suspect that doing `if [ "$OS" = "Windows_NT" ]` would trap all situations since that's defined by default on Windows, and wouldn't require any additional parameters. `SXA-ProcessCheck` was a bit of a hack so I wouldn't necessarily assume that its way of doing things is a good example of how to do things ;-)

File: terminateTestProcesses.sh (hunk 13 above)
Comment (andrew-m-leonard) [reply terminateTestProcesses.sh:52]: Ah ok! yes, so this and your other scripts will need altering for Sol11

File: terminateTestProcesses.sh [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+  
+  if [ $count -gt 0 ]; then
+      echo Windows rogue processes detected, attempting to stop them..
+      powershell -c "Get-WmiObject Win32_Process -Filter {${ignore_str} and ${match_str}}"
+      powershell -c "(Get-WmiObject Win32_Process -Filter {${ignore_str} and ${match_str}}).Terminate()"
+      uname | grep CYGWIN 2>/dev/null || uptime
+      echo Sleeping for 10 seconds...
+      sleep 10
+      uname | grep CYGWIN 2>/dev/null || uptime
```
Comment (sxa) [inline terminateTestProcesses.sh:33]: Suggest removing this line as it doesn't achieve anything as far as I can tell

File: terminateTestProcesses.sh (hunk 13 above)
Comment (andrew-m-leonard) [reply terminateTestProcesses.sh:50]: updated

**Review by sxa — APPROVED** [review]

**Comment by andrew-m-leonard** [comment]
@AdamBrousseau Hi Adam, if you could take a look as well please, as another CI user, thanks

**Comment by llxia** [comment]
zos [Grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/30138/console)

File: buildenv/jenkins/JenkinsfileBase [hunk 16]
```diff
@@ -829,50 +824,13 @@ def testExecution() {
...
+def terminateTestProcesses() {
+	echo "PROCESSCATCH: Terminating any hung/left over test processes:"
 
-	echo "PROCESSCATCH: Removed processes - here is what is remaining:"
-	if (PLATFORM.contains("windows")) {
-		sh(script:'powershell -c "Get-Process java | select id,processname,starttime"', returnStatus:true)
-	} else {
-		sh(script:'ps -fu '+"${env.USER}"+' | grep java | egrep -v "remoting.jar|agent.jar|grep"', returnStatus:true)
-	}
+        def statusCode=sh(script:"aqa-tests/terminateTestProcesses.sh ${env.USER} 2>&1", returnStatus:true)
```
Comment (karianna) [inline buildenv/jenkins/JenkinsfileBase:830]:
```suggestion
        def statusCode = sh(script:"aqa-tests/terminateTestProcesses.sh ${env.USER} 2>&1", returnStatus:true)
```

**Review by karianna — APPROVED** [review]
LGTM

**Review by llxia — APPROVED** [review]
LGTM

**Comment by AdamBrousseau** [comment]
Sorry I didn't get a chance to finish reviewing before merge. I am happy with it assuming it's been tested on all the platforms. We'll keep an eye out. We have a "sanitize" job we run a few times a day. This might allow us to disable that. Thanks!
